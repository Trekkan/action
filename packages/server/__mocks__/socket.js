/* eslint-env jest */
const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
}

export default socket
