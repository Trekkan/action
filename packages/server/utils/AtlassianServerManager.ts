import fetch from 'node-fetch'
import makeAppLink from './makeAppLink'
import AtlassianManager from 'parabol-client/utils/AtlassianManager'

interface AuthQueryParams {
  grant_type: 'authorization_code'
  code: string
  redirect_uri: string
}

interface RefreshQueryParams {
  grant_type: 'refresh_token'
  refresh_token: string
}

interface OAuth2Response {
  access_token: string
  refresh_token: string
  error: any
  scope: string
}

class AtlassianServerManager extends AtlassianManager {
  fetch = fetch
  static async init(code: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'authorization_code',
      code,
      redirect_uri: makeAppLink('auth/atlassian')
    })
  }

  static async refresh(refreshToken: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'refresh_token',
      refresh_token: refreshToken
    })
  }

  static async fetchToken(partialQueryParams: AuthQueryParams | RefreshQueryParams) {
    const queryParams = {
      ...partialQueryParams,
      client_id: process.env.ATLASSIAN_CLIENT_ID,
      client_secret: process.env.ATLASSIAN_CLIENT_SECRET
    }

    const uri = `https://auth.atlassian.com/oauth/token`

    const tokenRes = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(queryParams)
    })

    const tokenJson = (await tokenRes.json()) as OAuth2Response

    const {access_token: accessToken, refresh_token: refreshToken, error, scope} = tokenJson
    if (error) {
      throw new Error(`Atlassian: ${error}`)
    }
    const providedScope = scope.split(' ')
    const matchingScope =
      new Set([...AtlassianServerManager.SCOPE.split(' '), ...providedScope]).size ===
      providedScope.length
    if (!matchingScope) {
      throw new Error(`bad scope: ${scope}`)
    }
    return new AtlassianServerManager(accessToken, refreshToken)
  }

  refreshToken?: string

  constructor(accessToken: string, refreshToken?: string) {
    super(accessToken)
    this.refreshToken = refreshToken
  }
}

export default AtlassianServerManager
