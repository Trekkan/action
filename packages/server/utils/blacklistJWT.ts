import getBlacklistJWTKey from './getBlacklistJWTKey'
import {Threshold, SubscriptionChannel} from 'parabol-client/types/constEnums'
import {toEpochSeconds} from './epochTime'
import publish from './publish'
import getRedis from './getRedis'

const blacklistJWT = async (userId: string, iat: number, mutatorId?: string) => {
  const key = getBlacklistJWTKey(userId)
  const redis = getRedis()
  const expiresIn = iat - toEpochSeconds(new Date()) + Math.floor(Threshold.JWT_LIFESPAN / 1000)
  await redis.set(key, iat, 'EX', expiresIn)
  publish(SubscriptionChannel.NOTIFICATION, userId, 'InvalidateSessionsPayload', {}, {mutatorId})
}

export default blacklistJWT
