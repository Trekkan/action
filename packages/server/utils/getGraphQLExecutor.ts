import {ExecutionResult} from 'graphql'
import {ServerChannel} from 'parabol-client/types/constEnums'
import PubSubPromise from './PubSubPromise'

export interface PubSubPromiseResponse extends ExecutionResult {
  jobId: string
}

let pubsub: PubSubPromise<PubSubPromiseResponse>
const getGraphQLExecutor = () => {
  if (!pubsub) {
    pubsub = new PubSubPromise(
      ServerChannel.GQL_EXECUTOR_REQUEST,
      ServerChannel.GQL_EXECUTOR_RESPONSE
    )
  }
  return pubsub
}
export default getGraphQLExecutor
