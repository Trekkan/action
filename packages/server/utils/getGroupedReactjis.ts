import Reactji from '../database/types/Reactji'
import {IReactji} from 'parabol-client/types/graphql'

const getGroupedReactjis = (reactjis: Reactji[], viewerId: string, idPrefix: string) => {
  const agg = {}
  for (let i = 0; i < reactjis.length; i++) {
    const {id, userId} = reactjis[i]
    const guid = `${idPrefix}:${id}`
    const isViewerReactji = viewerId === userId
    const record = agg[guid]
    if (!record) {
      agg[guid] = {id: guid, count: 1, isViewerReactji}
    } else {
      record.count++
      record.isViewerReactji = record.isViewerReactji || isViewerReactji
    }
  }
  return Object.values(agg) as IReactji[]
}

export default getGroupedReactjis
