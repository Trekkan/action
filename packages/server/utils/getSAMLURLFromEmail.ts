import base64url from 'base64url'
import getSSODomainFromEmail from 'parabol-client/utils/getSSODomainFromEmail'
import getRethink from '../database/rethinkDriver'
import {URL} from 'url'

const urlWithRelayState = (url: string, isInvited?: boolean | null) => {
  if (!isInvited) return url
  const relayState = base64url.encode(JSON.stringify({isInvited: true}))
  const urlObj = new URL(url)
  urlObj.searchParams.append('RelayState', relayState)
  return urlObj.toString()
}

const getSAMLURLFromEmail = async (email: string, isInvited?: boolean | null) => {
  const domainName = getSSODomainFromEmail(email)
  if (!domainName) return null
  const r = await getRethink()
  const baseURL = (await r
    .table('SAML')
    .getAll(domainName, {index: 'domain'})
    .nth(0)('url')
    .default(null)
    .run()) as string | null
  if (!baseURL) return null
  return urlWithRelayState(baseURL, isInvited)
}

export default getSAMLURLFromEmail
