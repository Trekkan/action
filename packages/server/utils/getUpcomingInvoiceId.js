const getUpcomingInvoiceId = (orgId) => `upcoming_${orgId}`

export default getUpcomingInvoiceId
