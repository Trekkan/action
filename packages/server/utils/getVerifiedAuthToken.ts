import {verify} from 'jsonwebtoken'
import AuthToken from '../database/types/AuthToken'

const SERVER_SECRET_BUFFER = Buffer.from(process.env.SERVER_SECRET!, 'base64')

const getVerifiedAuthToken = (jwt: string | undefined | null, ignoreExp?: boolean) => {
  if (!jwt) return {} as AuthToken
  try {
    return verify(jwt, SERVER_SECRET_BUFFER, {ignoreExpiration: ignoreExp}) as AuthToken
  } catch (e) {
    return {} as AuthToken
  }
}

export default getVerifiedAuthToken
