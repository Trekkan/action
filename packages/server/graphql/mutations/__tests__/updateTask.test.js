/* eslint-env jest */
import getRethink from '../../../database/rethinkDriver'
import mockAuthToken from '../../../__tests__/setup/mockAuthToken'
import MockDate from 'mockdate'
import { __aMinuteAgo, __anHourAgo, __now } from '../../../__tests__/setup/mockTimes'
import fetchAndSerialize from '../../../__tests__/utils/fetchAndSerialize'
import DynamicSerializer from 'dynamic-serializer'
import MockDB from '../../../__tests__/setup/MockDB'
import socket from '../../../__mocks__/socket'
import { DONE } from 'parabol-client/utils/constants'
import convertToRichText from '../../../__tests__/setup/convertToRichText'
import updateTask from '../updateTask'
import makeDataLoader from '../../../__tests__/setup/makeDataLoader'

MockDate.set(__now)
console.error = jest.fn()

describe('updateTask', () => {
  test('updates the sortOrder without changing updatedAt or making a history item', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { task, user } = await mockDB.init().newTask()
    const taskId = task[0].id
    const authToken = mockAuthToken(user[7])
    const dataLoader = makeDataLoader(authToken)

    // TEST
    const updatedTask = {
      id: taskId,
      sortOrder: 2
    }
    await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })

    // VERIFY
    const db = await fetchAndSerialize(
      {
        task: r.table('Task').get(taskId),
        taskHistory: r
          .table('TaskHistory')
          .between([taskId, r.minval], [taskId, r.maxval], {
            index: 'taskIdUpdatedAt'
          })
          .orderBy({ index: 'taskIdUpdatedAt' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(dataLoader.isShared()).toEqual(true)
  })

  test('updates the content of the task', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { task, user } = await mockDB.init().newTask()
    const taskId = task[0].id
    const authToken = mockAuthToken(user[7])
    const dataLoader = makeDataLoader(authToken)
    // TEST
    const updatedTask = {
      id: taskId,
      content: convertToRichText('Updated content')
    }
    await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })

    // VERIFY
    const db = await fetchAndSerialize(
      {
        task: r.table('Task').get(taskId),
        taskHistory: r
          .table('TaskHistory')
          .between([taskId, r.minval], [taskId, r.maxval], {
            index: 'taskIdUpdatedAt'
          })
          .orderBy({ index: 'taskIdUpdatedAt' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(dataLoader.isShared()).toEqual(true)
  })

  test('updates the teamMember of the task slowly and trigger a new history item', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { task, teamMember, user } = await mockDB
      .init()
      .newTask({ updatedAt: new Date(__anHourAgo) })
      .newTaskHistory()
    const taskId = task[0].id
    const authToken = mockAuthToken(user[7])
    const dataLoader = makeDataLoader(authToken)

    // TEST
    const updatedTask = {
      id: taskId,
      teamMemberId: teamMember[5].id
    }
    await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })

    // VERIFY
    const db = await fetchAndSerialize(
      {
        task: r.table('Task').get(taskId),
        taskHistory: r
          .table('TaskHistory')
          .between([taskId, r.minval], [taskId, r.maxval], {
            index: 'taskIdUpdatedAt'
          })
          .orderBy({ index: 'taskIdUpdatedAt' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(dataLoader.isShared()).toEqual(true)
  })

  test('updates the content of the task quickly and do not trigger a new history item', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { task, user } = await mockDB
      .init()
      .newTask({ updatedAt: new Date(__aMinuteAgo) })
      .newTaskHistory()
    const taskId = task[0].id
    const authToken = mockAuthToken(user[7])
    const dataLoader = makeDataLoader(authToken)

    // TEST
    const updatedTask = {
      id: taskId,
      content: convertToRichText('Updated content')
    }
    await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })

    // VERIFY
    const db = await fetchAndSerialize(
      {
        task: r.table('Task').get(taskId),
        taskHistory: r
          .table('TaskHistory')
          .between([taskId, r.minval], [taskId, r.maxval], {
            index: 'taskIdUpdatedAt'
          })
          .orderBy({ index: 'taskIdUpdatedAt' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(dataLoader.isShared()).toEqual(true)
  })

  test('updates the status of the task', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { task, user } = await mockDB.init().newTask()
    const taskId = task[0].id
    const authToken = mockAuthToken(user[7])
    const dataLoader = makeDataLoader(authToken)

    // TEST
    const updatedTask = {
      id: taskId,
      status: DONE
    }
    await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })

    // VERIFY
    const db = await fetchAndSerialize(
      {
        task: r.table('Task').get(taskId),
        taskHistory: r
          .table('TaskHistory')
          .between([taskId, r.minval], [taskId, r.maxval], {
            index: 'taskIdUpdatedAt'
          })
          .orderBy({ index: 'taskIdUpdatedAt' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(dataLoader.isShared()).toEqual(true)
  })

  test('return error when the caller is not a team member', async () => {
    // SETUP
    const mockDB = new MockDB()
    const { task, user } = await mockDB.init().newTask()
    const authToken = mockAuthToken(user[1], { tms: ['foo'] })
    const taskId = task[0].id
    const dataLoader = makeDataLoader(authToken)
    // TEST
    const updatedTask = {
      id: taskId,
      status: DONE
    }
    const res = await updateTask.resolve(undefined, { updatedTask }, { authToken, dataLoader, socket })
    expect(res).toEqual(expect.objectContaining({ error: expect.any(Object) }))
  })
})
