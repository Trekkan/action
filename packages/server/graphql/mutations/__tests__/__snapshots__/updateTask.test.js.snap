// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateTask updates the content of the task 1`] = `
Object {
  "task": Object {
    "assigneeId": 3,
    "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Updated content\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "sortOrder": 0,
    "status": "active",
    "tags": Array [],
    "teamId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "taskHistory": Array [
    Object {
      "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Updated content\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
      "id": 4,
      "tags": Array [],
      "taskId": 0,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateTask updates the content of the task quickly and do not trigger a new history item 1`] = `
Object {
  "task": Object {
    "assigneeId": 3,
    "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Updated content\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "sortOrder": 0,
    "status": "active",
    "tags": Array [],
    "teamId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "taskHistory": Array [
    Object {
      "assigneeId": 3,
      "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Updated content\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
      "id": 4,
      "status": "active",
      "tags": Array [],
      "taskId": 0,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateTask updates the sortOrder without changing updatedAt or making a history item 1`] = `
Object {
  "task": Object {
    "assigneeId": 3,
    "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "sortOrder": 2,
    "status": "active",
    "tags": Array [],
    "teamId": 2,
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userId": 1,
  },
  "taskHistory": Array [],
}
`;

exports[`updateTask updates the status of the task 1`] = `
Object {
  "task": Object {
    "assigneeId": 3,
    "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "sortOrder": 0,
    "status": "done",
    "tags": Array [],
    "teamId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "taskHistory": Array [
    Object {
      "id": 4,
      "status": "done",
      "taskId": 0,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`updateTask updates the teamMember of the task slowly and trigger a new history item 1`] = `
Object {
  "task": Object {
    "assigneeId": 3,
    "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "sortOrder": 0,
    "status": "active",
    "tags": Array [],
    "teamId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
  "taskHistory": Array [
    Object {
      "assigneeId": 3,
      "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
      "id": 4,
      "status": "active",
      "taskId": 0,
      "updatedAt": 2017-01-01T07:00:00.000Z,
    },
    Object {
      "assigneeId": 3,
      "content": "{\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Task[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{}}",
      "id": 5,
      "status": "active",
      "taskId": 0,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;
